### PR TITLE
Added some ECS API sugar - new configurator type and a new place for us to define default systems when the time comes.

### LINKED ISSUE #351: Add ECS sugar for default systems.

### PR DESCRIPTION
**Pull request by RubyNova**
resolves #351

### CODE REVIEW

File: include/NovelRT/Maths/Utilities.h [hunk 1]
```diff
@@ -5,12 +5,19 @@
...
 
 #include <NovelRT/Exceptions/InvalidOperationException.h>
+#include <array>
 
 namespace NovelRT::Maths
 {
     class Utilities
     {
     public:
+        [[nodiscard]] inline constexpr static std::array<uint8_t, 32> Log2DeBruijin() noexcept
```
Comment (tannergooding) [inline include/NovelRT/Maths/Utilities.h:15]: Just calling out, before I forget, we'll want to add proper attribution here to both .NET Runtime and http://graphics.stanford.edu/~seander/bithacks.html

**Comment by RubyNova** [comment]
Please note this PR now also fixes the race condition introduced with a previous fix.

File: include/NovelRT/Maths/Utilities.h (hunk 1 above)
Comment (capnkenny) [reply include/NovelRT/Maths/Utilities.h:15]: So uhh...can I get just a quick explanation on what this is?

File: include/NovelRT/Ecs/SystemScheduler.h [hunk 2]
```diff
@@ -18,12 +18,6 @@ namespace NovelRT::Ecs
     class SystemScheduler
     {
     private:
-        struct QueueLockPair
-        {
-            std::optional<Atom> systemUpdateId;
-            std::mutex threadLock;
```
Comment (capnkenny) [inline include/NovelRT/Ecs/SystemScheduler.h:24]: No more locking required?

**Review by capnkenny — COMMENTED** [review]
Just need a few answers, tis all :)

File: src/NovelRT/Ecs/SystemScheduler.cpp [hunk 3]
```diff
@@ -57,111 +90,112 @@ namespace NovelRT::Ecs
...
-                auto& pair = _threadWorkQueues[workerIndex];
-
-                if (systemIds.empty())
-                {
-                    break;
-                }
+            uint64_t workerIndex;
 
-                const std::lock_guard<std::mutex> lock(pair.threadLock);
+            while ((workerIndex = NovelRT::Maths::Utilities::LeadingZeroCount64(_threadAvailabilityMap)) == 64)
```
[the remark this answers is not in the document]
Comment (RubyNova) [reply src/NovelRT/Ecs/SystemScheduler.cpp:121]: a `uint64_t` is 64 digits long - as such, if it returns 64, no workers are available.

File: include/NovelRT/Ecs/SystemScheduler.h (hunk 2 above)
Comment (RubyNova) [reply include/NovelRT/Ecs/SystemScheduler.h:24]: Yesn't.

We no longer lock in the traditional sense, but atomic operations might perform cheaper atomic locks on the XOR operations, or may just be volatile writes.

Comment (capnkenny) [reply include/NovelRT/Ecs/SystemScheduler.h:24]: So are we hoping that we hoping that we never lock in this case (lock-free)? Or do we want to just defer to the CPU to determine this at runtime?

Comment (RubyNova) [reply include/NovelRT/Ecs/SystemScheduler.h:24]: It's a case of the atomic lock would be so cheap on the XOR operations that it literally doesn't matter and you may as well consider this lockless due to the fact were using no mutexes at this point.

**Comment by capnkenny** [comment]
Just need an explanation of what tanner pointed out _as well as updated attribution_ before I shall approve

File: include/NovelRT/Maths/Utilities.h (hunk 1 above)
Comment (RubyNova) [reply include/NovelRT/Maths/Utilities.h:15]: Seems I spelt Debruijn wrong here.

**Comment by RubyNova** [comment]
In essence, the DeBruijn table provides a set of values for a table lookup which makes finding Log2 of the provided value faster as it takes less operations to perform.

In the future we'll be switching to the hardware intrinsic/asm instruction, but that is fiddly in C++, and so this software implementation was created.

**Pull request merged**